onnxruntime/core/providers/cpu/tensor/transpose.cc

```diff
@@ -25,72 +25,140 @@ size_t ComputeOffset(const std::vector<int64_t>& index, const std::vector<size_t
 
 // IncrementIndex: Increment an index into a tensor (in lexicographic ordering), wrapping
 // around the specified upper_bound.
-void IncrementIndex(std::vector<int64_t>& index, const std::vector<int64_t>& upper_bound, int64_t num_axes) {
+inline void IncrementIndex(std::vector<int64_t>& index, const std::vector<int64_t>& upper_bound, int64_t num_axes) {
   for (int64_t k = num_axes - 1; k >= 0; --k) {
     index[k]++;
     if (index[k] < upper_bound[k]) break;
     index[k] = 0;
   }
 }
 
+// DoTransposeSingleBlock: specialization of DoTranspose for the num_blocks=1 case.
+// copies source tensor to target, transposing elements.
+static inline void DoTransposeSingleBlock(size_t num_elts_in_block, const void* source, void* target,
+                                          size_t element_size) {
+  size_t blocksize = num_elts_in_block * element_size;
+  // copy
+  memcpy(target, source, blocksize);
+}
+
+static inline void DoTransposeSingleBlock(size_t num_elts_in_block, const std::string* source, std::string* target) {
+  const std::string* end = source + num_elts_in_block;
+  std::copy(source, end, target);
+}
```

Comment:
Is this going to call the std::string copy ctor for every entry? If so, do we need that given we're transposing?

Reply:
How would we avoid the copy, given that the original tensor may be freed (along with the strings it contains) while the transpose continues to be used? Without other memory management guarantees, it seems hard.

Reply:
Well, yeah, it is going to invoke operator=. I am not seeing how can we avoid that.

---
In reply to: [264969430](https://github.com/Microsoft/onnxruntime/pull/561#discussion_r264969430) [](ancestors = 264969430)

Reply:
You're right as usual. Can't avoid the copy here.

---
In reply to: [265253035](https://github.com/Microsoft/onnxruntime/pull/561#discussion_r265253035) [](ancestors = 265253035)